### ISSUE TITLE
Crash in BindingViewPagerAdapter

### ISSUE THREAD

**Issue by wongcain** [evant/binding-collection-adapter#148]
I'm finally getting around to upgrading to beta-3 in order to replace my own work-around for setting `LifecycleOwner` with yours.  I am getting a crash however in a Fragment that contains a ViewPager with a View that in turn has a RecyclerView.  I am using binding-collection-adapter for both the ViewPager as well as the RecyclerView.   The crash is occurring in `BindingViewPagerAdapter.instantiateItem(...)` when `binding.setLifecycleOwner(...)` is called.  I think the problem is that this is called before the `onBindBinding(...)` is called.  In my own implementation I am calling `binding.setLifecycleOwner(...)` inside the `onBindBinding(...)` method itself _after_ executing `binding.executePendingBindings()`.  I have found that the order of `executePendingBindings()` and `setLifecyclerOwner(...)` calls is important when dealing with nested data-bound views such as these.  For now I am reverting back to a previous version.  I will follow up on this issue with a Pull Request.

**Comment by evant** [comment]
Uploaded a version with a fix, can you try it out when you have a chance? 

```groovy
repositories {
  maven { url "https://dl.bintray.com/evant/maven" }
}
dependencies {
  compile 'me.tatarka.bindingcollectionadapter2:bindingcollectionadapter:3.0.0-beta4'
}
```
branch is: https://github.com/evant/binding-collection-adapter/tree/lifecycle-fix

**Comment by wongcain** [comment]
Unfortunately my project hasn't been upgraded to androidx yet.  Getting to that next month.   Can you upload the same for the backport?

**Comment by tonyatseek** [comment]
I have the same issue, and am using the back-port.

**Comment by 4h0q** [comment]
@evant it fixes the issue for me. I had the same problem with `beta3` though a slightly different case. The root cause was that `executePendingBindings` was called before the actual binding variable was set.

When will it be merged to the main branch & released?

thanks!

**Comment by evant** [comment]
3.0.0 has been released which includes this fix.

**Comment by 4h0q** [comment]
@evant Actually the issue is not fixed for my in `3.0.0`. It was fixed in commit `be54e475e4` though.

I think the issue with the current implementation is line 135 in:

https://github.com/evant/binding-collection-adapter/blob/1657869aae3ef48b5164f3aca4109cb2e6b8069b/bindingcollectionadapter/src/main/java/me/tatarka/bindingcollectionadapter2/BindingViewPagerAdapter.java:
`binding.setLifecycleOwner(lifecycleOwner);`

adding the lifecycle owner listener causes `executePendingBindings` to be "prematurely" called causing the crash.

**Comment by wongcain** [comment]
@evant  I think you need to switch the order of `binding.setLifeycleOwner(...)` and `binding.executePendingBindings()`.  This is counter-intuitive, but in my testing, this seems to be the correct order of operations.
In fact, I am currently using version `2.3.0-beta1` and I have extended your adapters with my own custom version.   In those I am doing the following:
```
    override fun onBindBinding(binding: ViewDataBinding, variableId: Int, layoutRes: Int,
                               position: Int, item: T) {
        super.onBindBinding(binding, variableId, layoutRes, position, item) // calls binding.executePendingBindings()
...
        if (lifecycleOwner != null) {
            binding.setLifecycleOwner(lifecycleOwner)
        }
    }```

These adapters are working without error.

**Comment by wongcain** [comment]
@evant Any update on this one?

**Comment by wongcain** [comment]
FYI - Google has now provided public a public getter for `LifecycleOwner` on `ViewDataBinding` so I don't think you need to use reflection any more.

**Comment by wongcain** [comment]
I've also found a problem with how you are attempting to set the `lifecycleOwner` in `BindingRecyclerViewAdapter`.  Specifically, you are attempting to do it in the method `onAttachedToRecyclerView`.  I have found that this is too early in the databinding lifecycle, and that the `lifecycleOwner` of the `RecyclerView`'s `ViewDataBinding` may be null here because it has not yet been set.   Instead I recommend trying to set it at the start of `onBindBinding`.

Below is a copy of my extended version of your `BindingRecyclerViewAdapter` and it is working perfectly.  (Note that I am still using version `2.3.0-beta1` of your library.)

```
open class LifecycleAwareBindingRecyclerViewAdapter<T> : BindingRecyclerViewAdapter<T>() {

    private var recyclerView: RecyclerView? = null
    private var lifecycleOwnerRef: WeakReference<LifecycleOwner>? = null
    
    /**
     * Sets the LifecycleOwner so that it can be set on item bindings.  Calling this method should
     * be unnecessary since the LifecycleOwner will be set automatically as long as the View 
     * containing the RecyclerView uses databinding and has its LifecycleOwner set.
     */
    fun setLifecycleOwner(lifecycleOwner: LifecycleOwner?) {
        if (lifecycleOwner != null) lifecycleOwnerRef = WeakReference(lifecycleOwner)
    }

    override fun onBindBinding(binding: ViewDataBinding, variableId: Int, layoutRes: Int,
                               position: Int, item: T) {
        trySetLifecycleOwner()
        super.onBindBinding(binding, variableId, layoutRes, position, item)
        var lifecycleOwner: LifecycleOwner? = lifecycleOwnerRef?.get()
        if (lifecycleOwner == null && binding.root.context is LifecycleOwner) {
            lifecycleOwner = binding.root.context as LifecycleOwner
        }
        if (lifecycleOwner != null) {
            binding.lifecycleOwner = lifecycleOwner
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        this.recyclerView = null
    }


    private fun trySetLifecycleOwner() {
        if (lifecycleOwnerRef?.get() == null) {
            val view = recyclerView ?: return
            val binding: ViewDataBinding = DataBindingUtil.findBinding(view) ?: return
            val bindingLifecycleOwner = binding.lifecycleOwner ?: return
            lifecycleOwnerRef = WeakReference(bindingLifecycleOwner)
        }
    }
}
```

**Comment by evant** [comment]
@wongcain thanks for the input. Feel free to submit a pr with those changes, otherwise I'll try to make them when I get a chance.

**Comment by justtwago** [comment]
@wongcain how do you use your `LifecycleAwareBindingRecyclerViewAdapter`? Where could I replace it with default `BindingRecyclerViewAdapter`?

**Comment by wongcain** [comment]
Create an instance of the adapter in your Activity or Fragment and set it as a variable in your data-bound layout.  Then just bind it to the RecyclerView like this:
```
            <android.support.v7.widget.RecyclerView
                android:layout_width="match_parent"
                android:layout_height="wrap_content"
                app:itemBinding="@{vm.itemsBinding}"
                app:items="@{vm.itemsList}"
                app:layoutManager="@{LayoutManagers.linear()}"
                **app:adapter="@{adapter}"**/>
```

**Comment by wongcain** [comment]
@evant I've finally got my project updated to AndroidX and have just issued a pull request to fix these crashes: https://github.com/evant/binding-collection-adapter/pull/160

Please review and let me know when you will be able to include these changes into a new release.
Thank you!

**Comment by evant** [comment]
Looks good! Will work on getting a release out in the coming week or so.

**Comment by wongcain** [comment]
Awesome.  Thanks for the quick response!

**Comment by evant** [comment]
Released as version `3.1.0`

**Comment by wongcain** [comment]
🎉
